local/network.go

```diff
@@ -435,59 +447,61 @@ func (ln *localNetwork) addNode(nodeConfig node.Config) (node.Node, error) {
 }
 
 // See network.Network
-func (ln *localNetwork) Healthy(ctx context.Context) chan error {
+func (ln *localNetwork) Healthy(ctx context.Context) error {
+	ln.lock.RLock()
+	defer ln.lock.RUnlock()
+
 	zap.L().Info("checking local network healthiness", zap.Int("nodes", len(ln.nodes)))
-	healthyChan := make(chan error, 1)
 
 	// Return unhealthy if the network is stopped
-	if ln.isStopped() {
-		healthyChan <- network.ErrStopped
-		return healthyChan
-	}
-
-	go func() {
-		// TODO: This will block the network for the duration of the health call.
-		// Maybe a better solution can be found.
-		ln.lock.RLock()
-		defer ln.lock.RUnlock()
-
-		errGr, cctx := errgroup.WithContext(ctx)
-		for _, node := range ln.nodes {
-			node := node
-			errGr.Go(func() error {
-				// Every constants.HealthCheckInterval, query node for health status.
-				// Do this until ctx timeout
-				for {
-					select {
-					case <-ln.closedOnStopCh:
-						return network.ErrStopped
-					case <-cctx.Done():
-						return fmt.Errorf("node %q failed to become healthy within timeout", node.GetName())
-					case <-time.After(healthCheckFreq):
-					}
-					health, err := node.client.HealthAPI().Health(cctx)
-					if err == nil && health.Healthy {
-						ln.log.Debug("node %q became healthy", node.name)
-						return nil
-					}
-				}
-			})
-		}
-		// Wait until all nodes are ready or timeout
-		if err := errGr.Wait(); err != nil {
-			healthyChan <- err
+	if ln.stopCalled() {
+		return network.ErrStopped
+	}
+
+	// Derive a new context that's cancelled when Stop is called,
+	// so that we calls to Healthy() below immediately return.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func(ctx context.Context) {
```

Comment:
does it need to receive ctx parameter? it is in the scope

Reply:
Yes, it needs to take `ctx` as an argument. Otherwise there's a race condition because the goroutine references `ctx` but `ctx` is re-assigned on line 475.